### PR TITLE
Support export/import plain keys for JDK8 in FIPS Mode

### PR DESCRIPTION
**Pull request by WilburZjh**
Signed-off-by: Jinhang Zhang <Jinhang.Zhang@ibm.com>

### CODE REVIEW

**Comment by WilburZjh** [comment]
@keithc-ca @pshipton @alon-sh

**Comment by WilburZjh** [comment]
ginder [cmdLineTester_fips](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder_FIPS/103/) test passed.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: jdk/src/share/classes/sun/security/pkcs11/P11Key.java [hunk 1]
```diff
@@ -367,6 +373,19 @@ static PrivateKey privateKey(Session session, long keyID, String algorithm,
             new CK_ATTRIBUTE(CKA_EXTRACTABLE),
         });
         if (attributes[1].getBoolean() || (attributes[2].getBoolean() == false)) {
+            if ((SunPKCS11.mysunpkcs11 != null) && "RSA".equals(algorithm)) {
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply jdk/src/share/classes/sun/security/pkcs11/P11Key.java:376]: added.

File: jdk/src/share/classes/sun/security/pkcs11/P11Key.java [hunk 2]
```diff
@@ -366,7 +372,20 @@ static PrivateKey privateKey(Session session, long keyID, String algorithm,
...
+                    RSAPrivateKey rsaPrivKey = RSAPrivateCrtKeyImpl.newKey(key);
+                    if (rsaPrivKey instanceof RSAPrivateCrtKeyImpl) {
+                        return new P11RSAPrivateKeyFIPS(session, keyID, algorithm, keyLength, attributes, (RSAPrivateCrtKeyImpl)rsaPrivKey);
+                    } else {
+                        return new P11RSAPrivateNonCRTKeyFIPS(session, keyID, algorithm, keyLength, attributes, rsaPrivKey);
+                    }
+                } catch (PKCS11Exception | InvalidKeyException e) {
+                    // Attempt failed, create a P11PrivateKey object.
+                }
+            }
```
Comment (keithc-ca) [inline jdk/src/share/classes/sun/security/pkcs11/P11Key.java:388]: The extra attribute check should only be for the FIPS case:
```
        if ((SunPKCS11.mysunpkcs11 != null) && "RSA".equals(algorithm)) {
            if (attributes[0].getBoolean() || attributes[1].getBoolean() || (attributes[2].getBoolean() == false)) {
                try {
                    byte[] key = SunPKCS11.mysunpkcs11.exportKey(session.id(), attributes, keyID);
                    RSAPrivateKey rsaPrivKey = RSAPrivateCrtKeyImpl.newKey(key);
                    if (rsaPrivKey instanceof RSAPrivateCrtKeyImpl) {
                        return new P11RSAPrivateKeyFIPS(session, keyID, algorithm, keyLength, attributes, (RSAPrivateCrtKeyImpl)rsaPrivKey);
                    } else {
                        return new P11RSAPrivateNonCRTKeyFIPS(session, keyID, algorithm, keyLength, attributes, rsaPrivKey);
                    }
                } catch (PKCS11Exception | InvalidKeyException e) {
                    // Attempt failed, create a P11PrivateKey object.
                }
            }
        }
        if (attributes[1].getBoolean() || (attributes[2].getBoolean() == false)) {
```

Comment (WilburZjh) [reply jdk/src/share/classes/sun/security/pkcs11/P11Key.java]: updated

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**